aiida/engine/processes/builder.py

```diff
@@ -238,6 +244,43 @@ def __init__(self, process_class: Type['Process']):
         self._process_spec = self._process_class.spec()
         super().__init__(self._process_spec.inputs)
 
+    def _repr_pretty_(self, p, _):  # pylint: disable=invalid-name
+        """Pretty representation for in the IPython console and notebooks."""
+
+        def format_inputs(inputs, level=0):
+            """Return the dictionary of inputs as a human-readable wayformatted multiline string for displaying.
+
+            :param inputs: a nested dictionary of inputs nodes
+            :param level: the nesting level of the current dictionary
+            """
+            result = []
+
+            def fmt_indent(content, level):
+                indent_string = ' ' * (4 * level)
+                return f'{indent_string}{content}'
+
+            for key, value in sorted(inputs.items()):
+                if isinstance(value, (dict, ProcessBuilder, ProcessBuilderNamespace)) and len(value) > 0:
+                    result.append(fmt_indent(f'"{key}": {{', level))
+                    result.append(format_inputs(value, level=level + 1))
+                    result.append(fmt_indent('},', level))
+                elif isinstance(value, Dict):
+                    result.append(fmt_indent(f'"{key}": {{', level))
+                    for line in json.dumps(value.get_dict(), indent=4).split('\n'):
+                        if line not in ['{', '}', '{}']:
+                            result.append(fmt_indent(f'{line}', level))
+                    result.append(fmt_indent('},', level))
+                elif isinstance(value, BaseType):
+                    result.append(fmt_indent(f'"{key}": {value.value}', level))
+                elif isinstance(value, Structure):
```

Comment:
We probably should not start doing "special cases" like this, and ideally use `Node.get_description`, which e.g. for `StructureData.get_description` returns `self.get_formula(mode='hill_compact')`

To be hypocritical, there is similar special-casing I added years back in: https://github.com/aiidateam/aiida-core/blob/2c5cc6fac2a1cea651d06089e44441a7e4e238fc/aiida/tools/visualization/graph.py#L204

but yeh both of these should really be standardised (I guess this ties in a bit with https://github.com/aiidateam/aiida-core/issues/3087)

Reply:
I agree that adding special cases is probably a bad idea long term. A bit confusing that there is both a `description` property and `get_description()` method, which return different strings.

I'm not sure if `get_description` should be the default fallback, or the `_repr_`. For a `Code`, you now get a good amount of information in one line:

```
Remote code 'qe-v6.7-pw' on localhost, pk: 1, uuid: 47a1c120-47af-4b9e-80c2-5a647c99817e
```

But if we switch to `get_description`, you get the description that was provided during code setup. That doesn't contain the PK, for example (similar issue for the `StructureData`, but that can be fixed).

Reply:
The fact that `Code.get_description` returns the value of the `description` property is just a fault in the `Code.get_description` implementation. That method is intended to give a rich description of the object. It is true that the name is not ideal and the implementation for the node types that ship with `aiida-core` are very inconsistent. But I agree with @chrisjsewell , we shouldn't compensate those errors by hard-coding special cases here. We should simply improve the `get_description` implementation of those node types.